[libc++] Document our ABI guarantees and what ABI flags exist to modify these guarantees (#132615)
llvm/llvm-project: libcxx/docs/ABIGuarantees.rst

.. _ABIGuarantees:

=======================
libc++'s ABI Guarantees
=======================

libc++ provides multiple types of ABI guarantees. These include stability of the layout of structs, the linking of TUs
built against different versions and configurations of the library, and more. This document describes what guarantees
libc++ provides in these different areas as well as what options exist for vendors to affect these guarantees.

Note that all of the guarantees listed below come with an asterisk that there may be circumstances where we deem it
worth it to break that guarantee. These breaks are communicated to vendors by CCing #libcxx-vendors on GitHub. If you
are a vendor, please ask to be added to that group to be notified about changes that potentially affect you.

ABI flags
=========
All the ABI flags listed below can be added to the ``__config_site`` header by the vendor to opt in to an ABI breaking

> change. These flags should never be set by the user. When porting libc++ to a new platform it should be considered by
> the vendor which flags should be enabled, assuming ABI stability is relevant to them. Please contact the libc++ team on

Reviewer: ```suggestion
change. These flags should never be set by the user. When porting libc++ to a new platform, vendors should consider which flags to enable, assuming that ABI stability is relevant to them. Please contact the libc++ team on
```

Discord or through other means to be able to make an informed decision on which flags make sense to enable, and to avoid
enabling flags which may not be stable. Flags can be enabled via the ``LIBCXX_ABI_DEFINES`` CMake option.

Stability of the Layout of Structs
==================================

> The layout of any user-observable struct is kept stable across versions of the library and any options users are allowed
> to change. There are a lot of structs that have internal names, but are none the less observable by users; for example

Reviewer: ```suggestion
The layout of any user-observable struct is kept stable across versions of the library and any user-facing options documented in https://libcxx.llvm.org/UserDocumentation.html#libc-configuration-macros. There are a lot of structs that have internal names, but are none the less observable by users; for example
```

This adds a bit of precision.

(see ``_LIBCPP_USING_DEV_RANDOM``). When switching from this implementation to another one on a platform that has
already shipped ``random_device``, one needs to retain the same object layout to remain ABI compatible. This flag
removes these workarounds for platforms that don't care about ABI compatibility.

``_LIBCPP_ABI_NO_COMPRESSED_PAIR_PADDING``
------------------------------------------
This removes artifical padding from ``_LIBCPP_COMPRESSED_PAIR`` and ``_LIBCPP_COMPRESSED_TRIPLE``. These macros are used
inside the associative and unordered containers, ``deque``, ``forward_list``, ``future``, ``list``, ``basic_string``,
``function``, ``shared_ptr``, ``unique_ptr``, and ``vector`` to stay ABI compatible with the legacy
``__compressed_pair`` type. ``__compressed_pair`` has historically been used to reduce storage requirements in the case
of empty types, but has been replaced by ``[[no_unique_address]]``. ``[[no_unique_address]]`` is significantly lighter
in terms of compile time and debug information, and also improves the layout of structs further. However, to keep ABI

> stability, the additional improvements in layout had to be reverted by introducing artificial padding.

Reviewer: ```suggestion
stability, the additional improvements in layout had to be reverted by introducing artificial padding. This setting removes that artificial padding.
```

Consider making it a separate paragraph after the first sentence?

These flags change the ``iterator`` member of various classes to reference hardened iterators instead. See the
:ref:`hardening documentation <hardening>` for more details.

Meaning of values
=================
The meaning of specific values can usually not be changed, since programs compiled against older versions of the headers
may check for these values. These specific values don't have to be hard-coded, but can also depend on user input.

There are multiple ABI flags that change the meaning of particular values:

``_LIBCPP_ABI_REGEX_CONSTANTS_NONZERO``
---------------------------------------
This changes the value of ``regex_constants::syntax_option-type::ECMAScript`` to be standards-conforming.

``_LIBCPP_ABI_FIX_CITYHASH_IMPLEMENTATION``
-------------------------------------------
This flag fixes the implementation of CityHash used for ``hash<fundamental-type>``. The incorrect implementation of
CityHash has the roblem that it drops some bits on the floor. Fixing the implementation changes the hash of values,
resulting in an ABI break.

inline namespaces
=================

> Inline namespaces which contain types that are ovservable by the user need to be kept the same, since they affect

Reviewer: ```suggestion
Inline namespaces which contain types that are observable by the user need to be kept the same, since they affect
```

> mangling. libc++ has almost all symbols inside an inline namespace. By default that namespace is ``__1``, but can be

Reviewer: ```suggestion
mangling. Almost all of libc++'s symbols are inside an inline namespace. By default that namespace is ``__1``, but can be
```
